driver-core/pom.xml

```diff
@@ -229,6 +229,8 @@
                             <instructions>
                                 <!-- JNR does not provide OSGi bundles, so exclude it; the driver can live without it -->
                                 <Import-Package><![CDATA[com.google.common.*;version="[16.0.1,22)",!jnr.*,*]]></Import-Package>
+                                <!-- Give the ability to discover and load netty-transport-native-epoll classes -->
+                                <DynamicImport-Package>*</DynamicImport-Package>
```

Comment:
I have limited knowledge of OSGI, but from what I can see in the [Netty PR](https://github.com/netty/netty/pull/5326/files), this is necessary to load native libraries. Is there any way we could target this more accurately? The `*` seems very generic (although I can't really tell the consequences).

Reply:
Native library is a corner case. DynamicImport-Package has the following main meaning: within OSGi container you have multiple bundles (JARs), each one is loaded by it's own classloader. OSGi controls visibility of java packages between bundles based on declaration in bundles manifests. Bundle can declare some packages as exported (Export-Package), it means that they can be used by other bundles. From consumer side - bundle declares what packages it needs using Import-Package. Usually import packages are generated automatically by maven-bungle-plugin based on imports in code. If a class is not a part of bundle itself, not a part of base JDK packages (in OSGI there is a boot delegation logic) and not declared as imported - it will not be resolved. 
There are cases when plain import-export logic is not enough - classes are not declared in code explicitly, the most common cause is Class.forName usage when class is loaded by name (as we have in com.datastax.driver.core.NettyUtil class). One possible way to deal with such dynamic class loading is to add DynamicImport-Package declaration. It allows to resolve class from another bundle if it is not declared explicitly as imported. Only exported packages can be imported via dynamic imports, so it does not break encapsulation.
Import-Package and DynamicImport-Package are also resolved at different moments of bundle lifecycle. Import-Package is resolved when a bundle is installed (and if a package is not present and not declared as optional - the resolving will fail). DynamicImport-Package is used in runtime, when an actual code tries to load classes and other ways do not work.
In theory we can try to limit DynamicImport-Package scope to  `io.netty.channel.epoll.*` but I did not test such option to be sure. Regarding consequences of using `*`  - actually there is no large difference because DynamicImport is used only as a last chance to load a class if other mechanisms do not work (the resolving order is a part of OSGi specification), so for existing working logic there is no large difference.